app/src/main/java/org/imozerov/streetartview/ui/ExploreArtActivity.kt

```diff
@@ -3,47 +3,88 @@ package org.imozerov.streetartview.ui
-import org.imozerov.streetartview.ui.helper.replaceFragment
+import org.imozerov.streetartview.ui.interfaces.Filterable
 import org.imozerov.streetartview.ui.map.ArtMapFragment
+import rx.android.schedulers.AndroidSchedulers
+import rx.subscriptions.CompositeSubscription
+import java.util.*
 
 class ExploreArtActivity : AppCompatActivity(), ArtObjectDetailOpener {
-    val MAIN_CONTENT_TAG = "MainContent"
+    val TAG = "ExploreArtActivity"
+
+    private var compositeSubscription : CompositeSubscription? = null
 
     override fun onCreate(savedInstanceState: Bundle?) {
+        Log.d(TAG, "onCreate()")
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_explore_art)
 
-        if (savedInstanceState == null) {
-            replaceMainContentWith(ArtListFragment.newInstance())
-        }
-
-        floating_button.setOnClickListener { v -> swapMainContent() }
+        val adapter = Adapter(supportFragmentManager)
+        adapter.addFragment(ArtListFragment.newInstance(), "List")
+        adapter.addFragment(ArtMapFragment.newInstance(), "Map")
+        viewpager.adapter = adapter
+        tabs.setupWithViewPager(viewpager)
     }
 
-    private fun swapMainContent() {
-        val currentFragment = supportFragmentManager.findFragmentByTag(MAIN_CONTENT_TAG)
-        when (currentFragment) {
-            is ArtListFragment -> replaceMainContentWith(ArtMapFragment.newInstance())
-            is ArtMapFragment -> replaceMainContentWith(ArtListFragment.newInstance())
-            else -> throw RuntimeException("Unknown fragment in main content! " + currentFragment);
-        }
+    override fun onStart() {
+        super.onStart()
+
+        val searchSubscription = RxSearchView
+                .queryTextChanges(search_view)
+                .doOnNext { Log.v(TAG, "Filtering $it") }
+                .subscribeOn(AndroidSchedulers.mainThread())
+                .subscribe {
+                    val fragment = supportFragmentManager
+                            .findFragmentByTag("android:switcher:" + R.id.viewpager + ":" + viewpager.currentItem)
+                    (fragment as? Filterable)?.applyFilter(it.toString())
+                }
+
+        compositeSubscription = CompositeSubscription();
+        compositeSubscription!!.add(searchSubscription)
     }
 
-    private fun replaceMainContentWith(fragment: Fragment) {
-        replaceFragment(R.id.main_content, fragment, MAIN_CONTENT_TAG)
+    override fun onStop() {
+        super.onStop()
+        compositeSubscription!!.clear()
     }
 
     override fun openArtObjectDetails(id: String?) {
+        Log.d(TAG, "openArtObjectDetails($id)")
         val intent = Intent(this, DetailArtObjectActivity::class.java)
         intent.putExtra(DetailArtObjectActivity.EXTRA_KEY_ART_OBJECT_DETAIL_ID, id)
         startActivity(intent)
     }
 
+    internal class Adapter(fm: FragmentManager) : FragmentPagerAdapter(fm) {
```

Comment:
When the file will grow we'll do so. Not now. For now the name is meaningful enough as it is internal class.